02-intermediate-ruby/references-and-values.md

```diff
@@ -0,0 +1,267 @@
+
+The second is a _reference_ to the array, which ties together the address of that memory with a name for our program to use. References are sometimes called _pointers_ (especially in C), and we say that a variable _points to_ or _references_ an object.
+
+![references and variables](images/references_and_values/references_and_values.png)
+
+Every variable in Ruby consists of these two parts, a reference and a value. Normally when you type the variable's name, Ruby automatically goes and gets the object. If you want to find out the address, you can use the `object_id` method:
+
+```ruby
+# You can see an object's memory address using object_id
+pets = ["dog", "parrot", "cat", "llama"]
+puts "pets.object_id: #{pets.object_id}"
+
+# Different objects have different IDs
+veggies = ["turnip", "beet"]
+puts "veggies.object_id: #{veggies.object_id}"
+```
+
+### The `=` Operator
+
+The `=` operator changes what a variable points at.
+
+If we assign one variable to another variable, they will both reference the same underlying object.
+
+```ruby
+# Two variables can point to the same object
+repeat = veggies
+puts "repeat.object_id: #{repeat.object_id}" # same as veggie.object_id
+```
+
+![referencing a variable twice](images/references_and_values/two_pointers.png)
+
+If we make changes to the object through one variable, you can see the changes via the other. The variables are just names, but the underlying object is the same.
+
+```ruby
+veggies[1] = "onion"
+repeat.push("potato")
+puts "#{veggies}"     # ["turnip", "onion", "potato"]
+puts "#{veggies}"     # ["turnip", "onion", "potato"]
+```
+
+![modifying the underlying object](images/references_and_values/modify_object.png)
+
+When we use the `=` operator, we are not changing the underlying object but instead _changing what our variable points to_. This does **not** affect any other variables.
+
+```ruby
+repeat = ["new", "array"]
+puts "repeat.object_id: #{repeat.object_id}"
+puts "value of repeat:"
+puts "#{repeat}"    # ["new", "array"]
+puts "value of veggies:"
+puts "#{veggies}"   # ["turnip", "onion", "potato"]
+```
+
+![creating a new array](images/references_and_values/new_array.png)
+
+So to summarize, if two variables point to the same underlying object:
+- Modifications to the _object_ (the value) will be visible from both variables
+- Reassigning one _variable_ (the reference) with `=` does not affect the other variable
+
+Note that `+=` and the other shorthand operators all involve reassignment. If we say `veggies += ['rutabaga']`, Ruby creates a new array, copies all the values from `veggies`, adds in `rutabaga`, and reassigns `veggies` to point to this new array. This is true of strings and numbers as well.
```

Comment:
I wonder if it might be good to highlight the difference between the `+=` and `<<` here? I can see that being a question that will come up since we tend to use `<<` more often in our curriculum